### PR TITLE
rgw/notifications: cache object size to avoid accessing invalid memoery

### PR DESCRIPTION
**Pull request by yuvalif**
Fixes: https://tracker.ceph.com/issues/51347

Signed-off-by: Yuval Lifshitz <ylifshit@redhat.com>

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by yuvalif** [comment]
valgrind issues are not observed in teuthology results:
http://pulpito.front.sepia.ceph.com/yuvalif-2021-07-02_16:07:30-rgw:verify-wip-yuval-test4-distro-basic-smithi/

* there are 4 `pg degraded` failures
* one failure in the following java test: `ObjectTest.testUploadFileList FAILED`

failures does not seem related to this fix. are they known issues?

**Comment by mkogan1** [comment]
> valgrind issues are not observed in teuthology results:
> http://pulpito.front.sepia.ceph.com/yuvalif-2021-07-02_16:07:30-rgw:verify-wip-yuval-test4-distro-basic-smithi/
> 
> * there are 4 `pg degraded` failures
> * one failure in the following java test: `ObjectTest.testUploadFileList FAILED`
> 
> failures does not seem related to this fix. are they known issues?

dont think that those issues are related 

but the strangely the valgrind issue still occurs sporadically - out of 2 teuthology runs on the same shaman build [1] the valgrind issue reproduced only once in a specific test:

_rgw:verify/{0-install centos_latest clusters/fixed-2 datacache/no_datacache frontend/beast ignore-pg-availability msgr-failures/few objectstore/bluestore-bitmap overrides proto/https rgw_pool_type/replicated sharding$/{single} striping$/{stripe-greater-than-chunk} tasks/{cls ragweed reshard s3tests-java s3tests} validater/valgrind}_ [2]

when `sharding$/{single}` but NOT when `sharding$/{default}`

_rgw:verify/{0-install centos_latest clusters/fixed-2 datacache/no_datacache frontend/beast ignore-pg-availability msgr-failures/few objectstore/bluestore-bitmap overrides proto/https rgw_pool_type/replicated sharding$/{default} striping$/{stripe-greater-than-chunk} tasks/{cls ragweed reshard s3tests-java s3tests} validater/valgrind}_ [3]


[1] https://shaman.ceph.com/builds/ceph/wip-rgw-d3n-cache-upstreaming_o23/ (d3n branch rebased on master + cherry-pick of the fix commit fc42652ff24828d029349c2c101015d2fdb4825a)

[2] https://pulpito.ceph.com/mkogan-2021-07-04_18:22:24-rgw:verify-wip-rgw-d3n-cache-upstreaming_o23-distro-basic-smithi/ --> 
https://pulpito.ceph.com/mkogan-2021-07-04_18:22:24-rgw:verify-wip-rgw-d3n-cache-upstreaming_o23-distro-basic-smithi/6252426/ -->
http://qa-proxy.ceph.com/teuthology/mkogan-2021-07-04_18:22:24-rgw:verify-wip-rgw-d3n-cache-upstreaming_o23-distro-basic-smithi/6252426/remote/smithi040/log/valgrind/ceph.client.0.log.gz
image
(d3n cache was disabled in this specific test per rgw log)

[3] https://pulpito.ceph.com/mkogan-2021-07-05_10:26:44-rgw:verify-wip-rgw-d3n-cache-upstreaming_o23-distro-basic-smithi/ -->
https://pulpito.ceph.com/mkogan-2021-07-05_10:26:44-rgw:verify-wip-rgw-d3n-cache-upstreaming_o23-distro-basic-smithi/6253107/

**Comment by yuvalif** [comment]
@mkogan1 do you see any of the valgrind issues that we had in the copy object operation?
errors I see are in the boost asio IO completion code?

**Comment by mkogan1** [comment]
> @mkogan1 do you see any of the valgrind issues that we had in the copy object operation?
> errors I see are in the boost asio IO completion code?

guess like they say _third time is the charm_ -
https://pulpito.ceph.com/mkogan-2021-07-05_13:53:56-rgw:verify-wip-rgw-d3n-cache-upstreaming_o23-distro-basic-smithi/ -->

the 

_rgw:verify/{0-install centos_latest clusters/fixed-2 datacache/no_datacache frontend/beast ignore-pg-availability msgr-failures/few objectstore/bluestore-bitmap overrides proto/https rgw_pool_type/replicated sharding$/{single} striping$/{stripe-greater-than-chunk} tasks/{cls ragweed reshard s3tests-java s3tests} validater/valgrind}_

test is GREEN:
https://pulpito.ceph.com/mkogan-2021-07-05_13:53:56-rgw:verify-wip-rgw-d3n-cache-upstreaming_o23-distro-basic-smithi/6253172/

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -5270,20 +5270,25 @@ void RGWCopyObj::execute(optional_yield y)
...
-    return;
-  }
 
-  if (!s->system_request) { // no quota enforcement for system requests
-    // enforce quota against the destination bucket owner
-    op_ret = dest_bucket->check_quota(this, user_quota, bucket_quota,
-				      astate->accounted_size, y);
+  uint64_t obj_size = 0;
+  {
+    // get src object size (cached in obj_ctx from verify_permission())
```
Comment (mkogan1) [inline src/rgw/rgw_op.cc:5276]: curious why the indentation, was this intended to be a `lock_guard()` block?

**Review by mkogan1 — APPROVED** [review]

File: src/rgw/rgw_op.cc (hunk 1 above)
Comment (cbodley) [reply src/rgw/rgw_op.cc:5276]: i like that it limits the scope of `RGWObjState* astate` so it can't be referenced after a possible invalidate

**Review by cbodley — APPROVED** [review]

**Pull request merged**